spec-abstract-noclosure/ffg.tla

```diff
@@ -0,0 +1,337 @@
+    IN
+    LET \* @type: ($vote, $vote) => Bool;
+        IsSurroundVoting(vote1, vote2) ==
+        /\ \/ vote1.ffg_vote.source[2] < vote2.ffg_vote.source[2]
+           \/ /\ vote1.ffg_vote.source[2] = vote2.ffg_vote.source[2]
+              /\ vote1.ffg_vote.source[1].slot < vote2.ffg_vote.source[1].slot
+        /\ vote2.ffg_vote.target[2] < vote1.ffg_vote.target[2]
+    IN
+    \E Slashable \in SmallQuorums:
+        \*/\ isMinCardinality(Cardinality(Slashable))
+        /\ \E vote1, vote2 \in votes:
+            /\ vote1.validator = vote2.validator
+            /\ vote1.validator \in Slashable
+            /\ IsEquivocation(vote1, vote2) \/ IsSurroundVoting(vote1, vote2)
+
+\* Append a block to chain 1.
+\* @type: Int => Bool;
+ProposeBlockOnChain1(slot) ==
+    LET new_block == Block(slot, NextBody(chain1_tip, "main")) IN
+    /\ IsValidBlock(new_block)
+    /\ slot > chain1_tip.slot
+    /\ all_blocks' = all_blocks \union { new_block }
+    /\ chain1' = chain1 \union { new_block }
+    /\ chain1_tip' = new_block
+    /\ IF IsForked
+       THEN UNCHANGED <<chain2, chain2_tip>>
+       ELSE
+         /\ chain2' = chain2 \union { new_block }
+         /\ chain2_tip' = new_block
+    /\ UNCHANGED <<ffg_votes, votes, justified_checkpoints, chain2_fork_block_number>>
+
+\* Append a block to chain 2.
+\* @type: Int => Bool;
+ProposeBlockOnChain2(slot) ==
+    LET new_block == Block(slot, NextBody(chain2_tip, "fork")) IN
+    /\ IsValidBlock(new_block)
+    /\ slot > chain2_tip.slot
+    /\ chain2_fork_block_number' =
+        IF IsForked THEN chain2_fork_block_number ELSE new_block.body
+    /\ all_blocks' = all_blocks \union { new_block }
+    /\ chain2' = chain2 \union { new_block }
+    /\ chain2_tip' = new_block
+    /\ UNCHANGED <<chain1, chain1_tip, ffg_votes, votes, justified_checkpoints>>
+
+JustifiedCheckpoints(viewVotes) ==
+    \* @type: Set($checkpoint) => Set($checkpoint);
+    LET AccJustified(justifiedSoFar, justifiedCheckpointSlot) ==
+        LET candidateCheckpoints == { Checkpoint(block, justifiedCheckpointSlot): block \in all_blocks } IN
+        LET newJustifiedCheckpoints == { c \in candidateCheckpoints: IsJustified(c, viewVotes, justifiedSoFar) } IN
+        justifiedSoFar \union newJustifiedCheckpoints
+    IN ApaFoldSeqLeft(AccJustified, { GenesisCheckpoint }, MkSeq(MAX_BLOCK_SLOT+2, (* @type: Int => Int; *) LAMBDA i: i))
+
+\* @type: ($checkpoint, $checkpoint, Set(Str)) => Bool;
+CastVotes(source, target, validators) ==
+    LET ffgVote == [ source |-> source, target |-> target ] IN
+    /\ IsValidFFGVote(ffgVote)
+    /\ validators /= {}
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    /\ LET allCheckpoints == { Checkpoint(block, i): block \in all_blocks, i \in CheckpointSlots } IN
+       \E allJustifiedCheckpoints \in SUBSET allCheckpoints:
+        /\ justified_checkpoints' = allJustifiedCheckpoints
```

Comment:
I think if we modify Lines 253 - 255 by the following constraints, it could be helpful for Apalache. While the new encoding requires a new variable and a new constraint, those may can reduce the search space.

```
/\ LET allCheckpoints == { Checkpoint(block, i): block \in all_blocks, i \in CheckpointSlots } 
           allValidCheckpoints == { c \in allCheckpoints :  IsValidCheckpoint(c) } 
   IN    \E allJustifiedCheckpoints \in SUBSET allValidCheckpoints:
              /\ justified_checkpoints' = allJustifiedCheckpoints
              /\ justified_checkpoints \subseteq justified_checkpoints'
```

Reply:
Good idea. It's not obvious whether it would help the current encoding, but let's give it a try.